### DISCUSSION TITLE
best practice to change default font family

### DISCUSSION THREAD

**Discussion by yoscbd** [holger1411/understrap#1797] [Q&A] [answered]
I', using bootstrap 5,
as i want to add a custom font as my website main font i wanted to ask what is the best practice to do so - where in the theme files should i make the change instead of just overriding the them default font family.

Thank you all!

**Comment by IanDelMar** [comment, marked as answer]
What are you trying to achieve? I'm not sure what is meant by "I want to change the _default_ font without overriding the _default_ font".

**Reply by yoscbd** [reply]
I want to set a custom font as the main font family for the theme,
I have notices that in **\wp-content\themes\understrap\src\sass\assets\bootstrap5\_variables.scss** the theme contain some variables that set the main font family....

My goal is to change the code in order to have my own custom font set in there and have it preloaded:

`$font-family-sans-serif: system-ui,
  -apple-system,
  "Segoe UI",
  Roboto,
  "Helvetica Neue",
  Arial,
  "Noto Sans",
  "Liberation Sans",
  sans-serif,
  "Apple Color Emoji",
  "Segoe UI Emoji",
  "Segoe UI Symbol",
  "Noto Color Emoji" !default;
$font-family-monospace: SFMono-Regular,
  Menlo,
  Monaco,
  Consolas,
  "Liberation Mono",
  "Courier New",
  monospace !default;

// stylelint-enable value-keyword-case
**$font-family-base: var(--#{$variable-prefix}font-sans-serif) !default;
$font-family-code: var(--#{$variable-prefix}font-monospace) !default;`**



I have also added this code in order to preload my custom font in :
**\wp-content\themes\understrap\inc\extras.php**


```
 add_action( 'wp_head', 'ybd_load_fonts' ); 
 function ybd_load_fonts() { 
	 ?> 

<link rel="preload" href="<?= get_template_directory_uri() ?>/fonts/RecifeTextWebTrial-Regular.woff2" as="font" type="font/woff2" crossorigin="anonymous" >
<link rel="preload" href="<?= get_template_directory_uri() ?>/fonts/FontsFree-Net-GT-America-Regular-1.woff2" as="font" type="font/woff2" crossorigin="anonymous" >

   <?php
}
```



Trying to figure out what is the best practice to add a custom font to the theme and have it preloaded as well.

Thank you!!!! :)

**Reply by IanDelMar** [reply]
You need to add a font-face declaration. (See https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face)

Add this to your _theme_variables.scss:
```
$font-path: "../fonts";

$font-family-sans-serif: GTAmerica, system-ui, -apple-system, "Segoe UI", Roboto, "Helvetica Neue", "Noto Sans", "Liberation Sans", Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";
$font-family-serif:  RecifeText;
```

Add this to the top of your _theme.scss:
```
@font-face {
  font-family: 'RecifeText';
  src: url('#{$font-path}/RecifeTextWebTrial-Regular.woff2');
  font-weight: normal;
  font-style: normal;
}

@font-face {
  font-family: 'GTAmerica';
  src: url('#{$font-path}/FontsFree-Net-GT-America-Regular-1.woff2');
  font-weight: normal;
  font-style: normal;
}
```

The font GT America will then be used as the default sans serif font. For the serif font you have to specify the CSS rules. E.g.
```
.myclass {
  font-family: $font-family-serif;
  ...
}
```

**Reply by yoscbd** [reply]
Thank you!
what about preloading the main font family?

before using font face to import the custom fonts i have this code that preloaded them as well:

```
add_action( 'wp_head', 'ybd_load_fonts' ); 
 function ybd_load_fonts() { 
	 ?> 

<link rel="preload" href="<?= get_template_directory_uri() ?>/fonts/RecifeTextWebTrial-Regular.woff2" as="font" type="font/woff2" crossorigin="anonymous" >
<link rel="preload" href="<?= get_template_directory_uri() ?>/fonts/FontsFree-Net-GT-America-Regular-1.woff2" as="font" type="font/woff2" crossorigin="anonymous" >
```

How can i archive the **preload** functionality when using font-face? 

**Should i simply do both?:**

**1. preloading the fonts:** 
`<link rel="preload" href="/fonts/RecifeTextWebTrial-Regular.woff2" as="font" type="font/woff2" crossorigin="anonymous" >
`

**2. adding font-face declaration with " font-display: swap" as you have shown me
Thank you again!**

**Reply by IanDelMar** [reply]
> what about preloading the main font family?

Preloading the font is a good idea and works the way you do it.

> How can i archive the preload functionality when using font-face? Should i simply do both?

Yes. The link tags tell the browser to preload the font files and the font-face declares the font. You have to use the font-face declaration. Preloading the font is good but optional.

**Comment by yoscbd** [comment]
Thank you for your much appreciated detailed answer ! :)
